### PR TITLE
[fpmsyncd] Fix memory leak bug

### PR DESCRIPTION
**Pull request by baiwei0427**
fpmsyncd: Fix memory leak bug

* Allocate and free Netlink memory in RouteSync::getIfName()

Signed-off-by: Wei Bai baiwei0427@gmail.com

### CODE REVIEW

File: fpmsyncd/routesync.cpp [hunk 1]
```diff
@@ -23,9 +23,7 @@ RouteSync::RouteSync(RedisPipeline *pipeline) :
     m_vnet_tunnelTable(pipeline, APP_VNET_RT_TUNNEL_TABLE_NAME, true),
     m_warmStartHelper(pipeline, &m_routeTable, APP_ROUTE_TABLE_NAME, "bgp", "bgp")
 {
-    m_nl_sock = nl_socket_alloc();
-    nl_connect(m_nl_sock, NETLINK_ROUTE);
-    rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
+
```
Comment (stcheng) [inline fpmsyncd/routesync.cpp]: remove empty line

File: fpmsyncd/routesync.cpp [hunk 2]
```diff
@@ -281,23 +279,67 @@ void RouteSync::onVnetRouteMsg(int nlmsg_type, struct nl_object *obj, string vne
...
+        nl_close(sk);
+        nl_socket_free(sk);
+        return false;
+    }
 
     /* Cannot get interface name. Possibly the interface gets re-created. */
-    if (!rtnl_link_i2name(m_link_cache, if_index, if_name, name_len))
+    if (!rtnl_link_i2name(cache, if_index, if_name, name_len))
     {
-        rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
```
Comment (15216401116) [inline fpmsyncd/routesync.cpp]: what about using` nl_cache_refill` method instead of  alloc and and free repeatedly?

**Comment by baiwei0427** [comment]
@jipanyang @stcheng  I guess the memory leak bug is due to the reallocation of link cache when interface name cannot be resolved. Is this bug first introduced on https://github.com/Azure/sonic-swss/commit/142673ad6cddf37286ab097c982b31f25c61358c#diff-4a3ea7752832ebd49f989cc95ae042e1?

File: fpmsyncd/routesync.cpp [hunk 3]
```diff
@@ -23,9 +23,6 @@ RouteSync::RouteSync(RedisPipeline *pipeline) :
     m_vnet_tunnelTable(pipeline, APP_VNET_RT_TUNNEL_TABLE_NAME, true),
     m_warmStartHelper(pipeline, &m_routeTable, APP_ROUTE_TABLE_NAME, "bgp", "bgp")
 {
-    m_nl_sock = nl_socket_alloc();
-    nl_connect(m_nl_sock, NETLINK_ROUTE);
-    rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
```
Comment (jipanyang) [inline fpmsyncd/routesync.cpp]: Could this netlink connection and cache allocation be kept here and reused at RouteSync::getIfName()?  So no need to  create netlink connection and alloc cache for each route message?

**Comment by stcheng** [comment]
@baiwei0427 yes, this pull request #61 is the place that introduced the function `rtnl_link_alloc_cache` in `onMsg` function. Thanks for narrowing down the issue.

One question I have here is that, since you replaced the function `rtnl_link_alloc_cache` with `nl_cache_refill`, do we need to re-allocate socket and cache every time we want to get the interface name, or could it be done during the initialization time and only do the `nl_cache_refill` in getIfName function?

In this case, the memory leak issue could also be addressed, is that correct?

File: fpmsyncd/routesync.cpp (hunk 3 above)
Comment (baiwei0427) [reply fpmsyncd/routesync.cpp]: @jipanyang This is a very good idea!

**Pull request merged**